File: 算法训练营/树/589/迭代法.py
#
# @lc app=leetcode.cn id=589 lang=python3
#
# [589] N叉树的前序遍历
#

# @lc code=start
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""


class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        if root is None:
            return []

        ret = []
        queue = [root]
        while queue:
            node = queue.pop()
            ret.append(node.val)
            if node.children:
                queue.extend(node.children[::-1])
        return ret
# @lc code=end

File: 算法训练营/20-字符串算法/917/迭代法.go
/*
 * @lc app=leetcode.cn id=917 lang=golang
 *
 * [917] 仅仅反转字母
 */

// @lc code=start
func reverseOnlyLetters(S string) string {
	bytes := []byte(S)
	start, end := 0, len(S)-1
	for ; start < end; start++ {
		if isAlpah(bytes[start]) { // 如果是字母
			for ;end > start; end-- {
				if isAlpah(bytes[end]) {
					break
				}
			}
			// 交换
			bytes[start], bytes[end] = bytes[end], bytes[start]
			end-- // 交换之后挪位
		}
	}
	return string(bytes)
}

func isAlpah(ch byte) bool {
	return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z')
}
// @lc code=end